src/crimson/osd/osd.cc

```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
+          shard_services.start_operation<ScrubEvent>(
+            pgref, *osds, spg,
+            (ScrubEvent::ScrubEventFwdImm)(&PgScrubber::scrub_echo_v),
+            pgref->get_osdmap_epoch(),
+            ++dbg_idx+100, 100ms);
+           return seastar::now();
+        });
+    }
+  }
+  logger().warn("scrub_tick(): after pg loop");
+#endif
 }
+
+
+seastar::future<> OSD::sched_scrub()
+{
+  if (!state.is_active()) {
+    return seastar::now();
+  }
+  if (scrub_random_backoff()) {
+    return seastar::now();
+  }
+
+  // fail fast if no resources are available
+  if (!scrub_scheduler.can_inc_scrubs()) {
+    logger().info("{}: OSD cannot incr scrub", __func__);
+    return seastar::now();
+  }
+
+  // if there is a PG that is just now trying to reserve scrub replica resources
+  // - we should wait and not initiate a new scrub
+  if (scrub_scheduler.is_reserving_now()) {
+    logger().info("{}: scrub resources reservation in progress", __func__);
+    return seastar::now();
+  }
+
+  // find something to scrub
+
+  Scrub::ScrubPreconds env_conditions;
+
+  if (shard_services.is_recovery_active() &&
+      !local_conf()->osd_scrub_during_recovery) {
+    if (!local_conf()->osd_repair_during_recovery) {
+      logger().info("{}: not scheduling scrubs due to active recovery",
+                    __func__);
+      return seastar::now();
+    }
+    logger().info(
+      "OSD::sched_scrub(): scheduling explicitly requested repair due to "
+      "active recovery");
+    env_conditions.allow_requested_repair_only = true;
+  }
+
+  if (g_conf()->subsys.should_gather<ceph_subsys_osd, 20>()) {
+    logger().debug("{}: scrub_sched_start", __func__);
+    auto all_jobs = scrub_scheduler.list_registered_jobs();
+    for (const auto& sj : all_jobs) {
+      logger().debug("OSD::sched_scrub(): scrub-queue jobs: {}", *sj);
+    }
+  }
+
+  return scrub_scheduler.select_pg_and_scrub(std::move(env_conditions))
+    .then([](auto was_started) {
+      logger().info("OSD::sched_scrub(): sched_scrub done ({})",
+                    ScrubQueue::attempt_res_text(was_started));
+      return seastar::now();
+    });
+}
+
+
+bool OSD::scrub_random_backoff()
+{
+  bool coin_flip =
+    (rand() / (double)RAND_MAX) >= local_conf()->osd_scrub_backoff_ratio;
+
+  /* RRR for now */ coin_flip = true;
+  if (!coin_flip) {
+    logger().info("scrub_random_backoff lost coin flip, randomly backing off");
+    return true;
+  }
+  return false;
+}
+
+
+seastar::future<Scrub::schedule_result_t> OSD::initiate_a_scrub(
+  spg_t pgid, bool allow_requested_repair_only)
+{
+  logger().info("{}: trying to initiate a scrubbing of {}", __func__, pgid);
+
+  // we have a candidate to scrub. We need some PG information to know if
+  // scrubbing is allowed
+
+  Ref<PG> pg = get_pg(pgid);
+  if (!pg) {
+    // the PG was dequeued in the short timespan between creating the candidates
+    // list (collect_ripe_jobs()) and here
```

Comment:
Would be a way to express this as a machine validated assertion?

Reply:
That would be wrong. PG's do come and go.